src/main/scala/viper/gobra/frontend/Parser.scala

```diff
@@ -307,1232 +262,96 @@ object Parser {
-
-    lazy val ghostStatement: Parser[PGhostStatement] =
-      "ghost" ~> statement ^^ PExplicitGhostStatement |
-      "assert" ~> expression ^^ PAssert |
-      "exhale" ~> expression ^^ PExhale |
-      "assume" ~> expression ^^ PAssume |
-      "inhale" ~> expression ^^ PInhale |
-      "fold" ~> predicateAccess ^^ PFold |
-      "unfold" ~> predicateAccess ^^ PUnfold
-
-    lazy val ghostParameter: Parser[Vector[PParameter]] = {
-      val namedParam =
-        "ghost" ~> rep1sep(idnDef, ",") ~ "...".? ~ typ ^^ {
-          case ids ~ variadicOpt ~ t => ids map { id =>
-            val typ = if (variadicOpt.isDefined) PVariadicType(t.copy) else t.copy
-            PExplicitGhostParameter(PNamedParameter(id, typ).at(id)).at(id)
+    def parse(rule : => Rule): Either[Vector[ParserError], Node] = {
+      val name = source.name
+      val tree = try time ("ANTLR_PARSE_SLL", name) { rule }
+      catch {
+        case _: AmbiguityException  => parse_LL(rule) // Resolve `<IDENTIFIER> { }` ambiguities in switch/if-statements
+        case _: ParseCancellationException => parse_LL(rule) // For even faster parsing, replace with `new ParserRuleContext()`.
+        case e => errors.append(ParserError(e.getMessage, Some(SourcePosition(source.toPath, 0, 0)))); new ParserRuleContext()
+      }
+      if(errors.isEmpty) {
+        val translator = new ParseTreeTranslator(pom, source, specOnly)
+        val parseAst : Node = time("ANTLR_TRANSLATE", source.name) {
+          try translator.translate(tree)
+          catch {
+            case e: TranslationFailure =>
+              val pos = source match {
+                case fileSource: FromFileSource => Some(SourcePosition(fileSource.path , e.cause.startPos.line, e.cause.endPos.column))
+                case _ => None
+              }
+              return Left(Vector(ParserError(e.getMessage + " " + e.getStackTrace.toVector(1), pos)))
+            case e : UnsupportedOperatorException =>
+              val pos = source match {
+                case fileSource: FromFileSource => Some(SourcePosition(fileSource.path, e.cause.startPos.line, e.cause.endPos.column))
+                case _ => None
+              }
+              return Left(Vector(ParserError(e.getMessage + " " + e.getStackTrace.toVector(0), pos)))
+            case e =>
+              val pos = source match {
+                case fileSource: FromFileSource => Some(SourcePosition(fileSource.path, 0, 0))
+                case _ => None
+              }
+              return Left(Vector(ParserError(e.getMessage + " " + e.getStackTrace.take(4).mkString("Array(", ", ", ")"), pos)))
           }
         }
-
-      val unnamedParam =
-        "ghost" ~> "...".? ~ typ ^^ {
-          case variadicOpt ~ t =>
-            val typ = if (variadicOpt.isDefined) PVariadicType(t) else t
-            Vector(PExplicitGhostParameter(PUnnamedParameter(typ).at(t)).at(t))
+        // TODO : Properly handle parser warnings
```

Comment:
what kind of warnings are this? If the ANTLR grammar has a mistake?

Reply:
If that is the case, then this is ok. Otherwise, please be clearer regarding what has to be changed.

Reply:
In some cases when Gobra allows some structure that is not allowed in Go, like struct literals that would have to be parenthesised, I create a warning in the translator. We decided not to emit errors, since it's not always detected, so it would be a bit inconsistent. I think Viper verification results don't support warnings right? Either everything is okay or it is a failure with errors?

Reply:
Just clarify the "Properly handle parser warning" so that we know what is meant by this in the future.